### PR TITLE
[Observer] Triton grid search for the MSE observer

### PR DESCRIPTION
**Pull request by Bias92**
## Summary

Adds a triton kernel for MSE observer

- remove compile path
- added a triton_error_buffer that works with patience. since the triton programs are per group, the native patience metric is much quicker to end the search compared to the eager mode's whole tensor patience. This makes accuracy much worse. To compensate we either have to get rid of patience for triton kernel [slow] or modify it so it is less aggresive. a 30% buffer seems to give approximately similar error/evaluation number characteristics. I tried various different patience numbers to align eager/triton behavior (e.g. 3x patience for triton or something like that) but it seems to behave a lot differently, resulting in more evaluations but worse accuracy unless the patience is set ridiculously high, in which case its just a lot more evaluations.
- triton kernel achieves bitwise parity with eager mode when patience/buffer are set to do full evaluation

benchmarks and kernel development: https://gist.github.com/HDCharles/51c392795d109078cb89b2f45ccfa11c

TODO: unify triton helpers between CT/LLMC

### Kernel details

The kernel uses buffered per-qparam patience and adaptive 512-value tiling:

- Small groups are packed so each Triton program handles roughly 512 values.
- Large groups are split into 512-value chunks, followed by a partial-error reduction.
- Candidate qparam calculation and QDQ are JIT-inlined helpers in `triton_utils.py`.
- INT, FP4, FP8, FP16/BF16/FP8 scale rounding, and E8M0 scales are supported.
- `triton_error_buffer` defaults to 100% for FP4 and 30% otherwise.
- The previous compile path and observer `chunk_size` setting are removed.

## Numerical parity

The packed BF16 NVFP4 path now reproduces eager arithmetic exactly, including:

- observed-dtype normalization before FP4 midpoint rounding;
- observed-dtype candidate min/max and scale rounding;
- observed-dtype `pow` inputs/results with FP32 reduction accumulation;
- host-precomputed FP32 grid points, avoiding reciprocal-rounding differences.

The deterministic expanded-NVFP4 Llama-3.1-8B output is bitwise identical to main. 

The exact-parity changes cost approximately 9-11% relative to the original adaptive prototype:

| Weight shape | Prototype | Parity kernel | Overhead |
|---|---:|---:|---:|
| Q/O projection | 8.324 ms | 9.100 ms | 9.3% |
| K/V projection | 2.333 ms | 2.579 ms | 10.6% |
| Up/Gate projection | 28.225 ms | 30.839 ms | 9.3% |
| Down projection | 28.230 ms | 30.860 ms | 9.3% |

for `meta-llama/Meta-Llama-3.1-8B-Instruct`

## End-to-end results

Expanded NVFP4 MSE example, `meta-llama/Meta-Llama-3.1-8B-Instruct`:

| Metric | Main eager | Triton parity | Speedup |
|---|---:|---:|---:|
| Full runtime | 335.65 s | 211.16 s | 1.59x |
| MSE grid search | 39.40 s | 3.73 s | 10.56x |
| Peak GPU allocated | 2.66 GB | 2.66 GB | same |
| Peak GPU reserved | 3.48 GB | 3.24 GB | 6.8% lower |

## Tests

Observer tests cover eager/Triton exact matching across INT4, FP4, FP8, tensor, channel, group, tensor-group and block layouts. A dedicated BF16 NVFP4 group-size-16 regression test exercises the packed path and its FP4 midpoint behavior.

```text
21 passed
```

### CODE REVIEW

File: src/llmcompressor/observers/mse_triton.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    # would be uninitialised memory. Eager lands on the same values there,
+    # since its search loop does not execute either.
+    best_min = min_val.clone()
+    best_max = max_val.clone()
+    for i in range(total_steps):
+        won = best_step == i
+        p = 1.0 - i / grid
+        best_min = torch.where(won, min_val * p, best_min)
+        best_max = torch.where(won, max_val * p, best_max)
+    return best_min, best_max
```
[the remark this answers is not in the document]
Comment (Bias92) [reply src/llmcompressor/observers/mse_triton.py]: i measured this one before taking it. building `ps` in `min_val.dtype` rounds `p` before the multiply, and eager multiplies by a python float, so the two only agree in fp32. over 131072 values it changes 26543 of them in bf16 and 25291 in fp16, up to 7.8e-03 relative.

thats the same class of thing the kernel already had to fix. the scale has to be rebuilt per step in the observed dtype rather than scaled once in fp32, which was worth 345 of 2048 groups on bf16 int8. keeping the loop here for the same reason.

the allocation is one tensor either way, `empty_like` vs `clone`, and this runs once per observer call rather than per step, so the loop is not on a hot path.

File: src/llmcompressor/observers/mse_triton.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+      there would be silently dropped.
+    - the group has to fit a block the kernel has been run at.
+    - CUDA proper, at compute capability 8.0 or above. torch.Tensor.is_cuda
+      is also True under ROCm, the error path calls into libdevice, and
+      Triton does not support older NVIDIA architectures.
+    """
+    return (
+        _TRITON_AVAILABLE
+        and observed.is_cuda
+        and torch.version.hip is None
```
[the remark this answers is not in the document]
Comment (Bias92) [reply src/llmcompressor/observers/mse_triton.py]: took this one, thanks. applied in a8c1ea26.

**Review by kylesayrs — APPROVED** [review]
